chips/msp432/src/chip.rs

```diff
@@ -0,0 +1,100 @@
+use core::fmt::Write;
+use cortexm4;
+use kernel::Chip;
+
+use crate::gpio;
+use crate::nvic;
+use crate::timer;
+use crate::uart;
+use crate::wdt;
+
+pub struct Msp432 {
+    mpu: cortexm4::mpu::MPU,
+    userspace_kernel_boundary: cortexm4::syscall::SysCall,
+    scheduler_timer: cortexm4::systick::SysTick,
+}
+
+impl Msp432 {
+    pub unsafe fn new() -> Msp432 {
+        Msp432 {
+            mpu: cortexm4::mpu::MPU::new(),
+            userspace_kernel_boundary: cortexm4::syscall::SysCall::new(),
+            scheduler_timer: cortexm4::systick::SysTick::new_with_calibration(48_000_000),
+        }
+    }
+}
+
+impl Chip for Msp432 {
+    type MPU = cortexm4::mpu::MPU;
+    type UserspaceKernelBoundary = cortexm4::syscall::SysCall;
+    type SchedulerTimer = cortexm4::systick::SysTick;
+    type WatchDog = wdt::Wdt;
+
+    fn service_pending_interrupts(&self) {
+        unsafe {
+            loop {
+                if let Some(interrupt) = cortexm4::nvic::next_pending() {
+                    match interrupt {
+                        nvic::USCI_A0 => uart::UART0.handle_interrupt(),
+                        nvic::IO_PORT1 => gpio::handle_interrupt(0),
+                        nvic::IO_PORT2 => gpio::handle_interrupt(1),
+                        nvic::IO_PORT3 => gpio::handle_interrupt(2),
+                        nvic::IO_PORT4 => gpio::handle_interrupt(3),
+                        nvic::IO_PORT5 => gpio::handle_interrupt(4),
+                        nvic::IO_PORT6 => gpio::handle_interrupt(5),
+                        nvic::TIMER_A0_0 | nvic::TIMER_A0_1 => timer::TIMER_A0.handle_interrupt(),
+                        nvic::TIMER_A1_0 | nvic::TIMER_A1_1 => timer::TIMER_A1.handle_interrupt(),
+                        nvic::TIMER_A2_0 | nvic::TIMER_A2_1 => timer::TIMER_A2.handle_interrupt(),
+                        nvic::TIMER_A3_0 | nvic::TIMER_A3_1 => timer::TIMER_A3.handle_interrupt(),
+                        _ => {
+                            panic!("unhandled interrupt {}", interrupt);
+                        }
+                    }
+
+                    let n = cortexm4::nvic::Nvic::new(interrupt);
+                    n.clear_pending();
+                    n.enable();
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn has_pending_interrupts(&self) -> bool {
+        unsafe { cortexm4::nvic::has_pending() }
+    }
+
+    fn mpu(&self) -> &cortexm4::mpu::MPU {
+        &self.mpu
+    }
+
+    fn scheduler_timer(&self) -> &cortexm4::systick::SysTick {
+        &self.scheduler_timer
+    }
+
+    fn watchdog(&self) -> &Self::WatchDog {
+        unsafe { &wdt::WDT }
```

Comment:
This is concerning, why is it unsafe to use the watchdog here?

Reply:
Because the rust-compiler won't let me return a reference to a static mutable variable. Because there is only 1 watchdog instances I thought this should not be a big problem although it is marked as unsafe.

Reply:
I think that rather than initializing the watchdog timer as a global, it should just be a field of the MSP432 struct, and be initialized in the constructor for that struct.

Using it this way opens the door for future code in the kernel to take a simultaneous mutable reference to the WDT, violating Rust's aliasing guarantees.

Reply:
Ahh, I see!
Then I create a field within the Msp432 struct for the watchdog and delete the global instance.

Reply:
Is it necessary to stop it immediately at startup? Could that just be done as part of the call to `msp432::new()`? Either way, I think that temporarily allocating a short lived instance on the stack (as you are doing) is also fine.

Reply:
I think it is necessary to stop it immediately at startup, because the interval for it is only ~10ms and depending on what is done before the Msp432 struct is instantiated it could be that the watchdog resets the system.